README.md
## Usage

Must(Initialize from the start page:Your Home Page)
```dart
  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addPostFrameCallback((_) {
      FlutterLogOverlay.init(context: context);
    });
  }
```

Use
```dart
FlutterLogOverlay.show(); 
```

plugin for logger
restructure Logger
```dart
final sLog = MyLogger(
  printer: PrettyPrinter(
    methodCount: 0,
    errorMethodCount: 0,
    printEmojis: false,
    printTime: true,
  ),
);

class MyLogger extends Logger {
  LogFilter? filter;
  LogPrinter? printer;
  LogOutput? output;
  Level? level;

  MyLogger({
    this.filter,
    this.printer,
    this.output,
    this.level,
  }) : super(
          filter: filter,
          printer: printer,
          output: output,
          level: level,
        );

  @override
  void log(
    Level level,
    dynamic message, [
    dynamic error,
    StackTrace? stackTrace,
  ]) {
    _startTime = DateTime.now();
    FlutterLogOverlay.addLog(
      isCore: level == Level.error,
      content: [
        getTime(),
        "--------------------",
        ((printer ?? PrettyPrinter()) as PrettyPrinter)
            .stringifyMessage(message),
      ],
    );
    super.log(
      level,
      message,
      error != null || stackTrace != null
          ? [
              error,
              stackTrace,
            ]
          : null,
    );
  }

  DateTime? _startTime;

  String getTime() {
    var now = DateTime.now();
    var year = _fourDigits(now.year);
    var month = _twoDigits(now.month);
    var day = _twoDigits(now.day);
    var h = _twoDigits(now.hour);
    var min = _twoDigits(now.minute);
    var sec = _twoDigits(now.second);
    var ms = _threeDigits(now.millisecond);
    var timeSinceStart = now.difference(_startTime!).toString();
    return '$year-$month-$day $h:$min:$sec.$ms (+$timeSinceStart)';
  }

  String _fourDigits(int n) {
    if (n >= 1000) return '$n';
    if (n >= 100) return '0$n';
    if (n >= 10) return '00$n';
    return '000$n';
  }

  String _threeDigits(int n) {
    if (n >= 100) return '$n';
    if (n >= 10) return '0$n';
    return '00$n';
  }

  String _twoDigits(int n) {
    if (n >= 10) return '$n';
    return '0$n';
  }
}
```

Optional
```dart
  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onPanEnd: FlutterLogOverlay.showOverlayPan,
      behavior: HitTestBehavior.translucent,
      child: const MaterialApp(
        debugShowCheckedModeBanner: false,
        home: FirstPgae(),
      ),
    );
  }
```

CHANGELOG.md
# 0.0.1

* overlay windows for log

## 0.0.2

* update LICENSE

## 0.0.3

* update LICENSE

## 0.0.4

* update homepage

## 0.0.5

* update for restructure Logger from plugin:logger

## 0.0.6

* update for restructure Logger from plugin:logger

## 0.0.7

* update LICENSE and 'pubspec.yaml'

## 0.0.8

* log ListView refresh from bottom

## 0.0.9

* add stow function,log TextStyle and scroll bar

## 0.0.10

* fix the deployment placeholder problem
  修复展开时占位问题
* optimize rendering logic
  优化渲染逻辑

## 0.0.11

* update Text to SelectableText
  Text替换SelectableText


lib/src/log_overlay.dart
import 'package:flutter/material.dart';

import 'log_overlay_model.dart';
import 'log_overlay_widget.dart';

class FlutterLogOverlay {
  ///init:需要获取home的context
  ///
  ///(MaterialApp的context还未对页面开始初始化，无法获取MediaQuery)
  ///
  ///(需要在home的initState中第二帧开始初始化)
  static late BuildContext _currentContext;

  ///init:弹窗的宽度
  static double _width = 0.0;

  ///init:弹窗的高度
  static double _height = 0.0;

  ///init:控制弹窗的左上角的x轴坐标(相对于屏幕)
  static double _overlayLeft = 0.0;

  ///init:控制弹窗的左上角的yx轴坐标(相对于屏幕)
  static double _overlayTop = 0.0;

  ///init:标题的高度
  static double _titleHeight = 0.0;

  ///init:标题的背景颜色
  static Color? _barColor;

  ///init:消息的背景颜色
  static Color? _itemColor;

  ///init:error消息的背景颜色
  static Color? _errorColor;

  ///init:需要标题
  static bool _needTitle = true;

  ///init:需要清除按钮
  static bool _needClean = true;

  ///init:日志的TextStyle
  static TextStyle? _logTextStyle;

  ///悬浮是否显示
  static bool get isShow => entry.mounted;

  ///是否展开listview
  static bool showListView = true;

  ///可自定义widget(LogOverlayWidget)
  static Widget? logOverlayWidget;

  ///日志缓存
  static final List<LogOverlayModel> logList = [];

  ///context的mediaQueryData
  static final _mediaQueryData = MediaQuery.of(_currentContext);

  ///context的overlay
  static final _overlay = Overlay.of(_currentContext);

  ///初始化
  ///
  ///[context]为当前页面的context
  static void init({
    required BuildContext context,
    double? width,
    double? height,
    double? overlayLeft,
    double? overlayTop,
    double? titleHeight,
    Color? backgroundColor,
    Color? barColor,
    Color? itemColor,
    Color? errorColor,
    bool needTitle = true,
    bool needClean = true,
    TextStyle? logTextStyle,
  }) {
    _currentContext = context;
    _width = width ?? 300.0;
    _height = height ?? 500.0;
    _overlayLeft = overlayLeft ?? _mediaQueryData.padding.left;
    _overlayTop = overlayTop ?? _mediaQueryData.padding.top;
    _titleHeight = titleHeight ?? 60.0;
    _barColor = barColor;
    _itemColor = itemColor;
    _errorColor = errorColor;
    _needTitle = needTitle;
    _needClean = needClean;
    _logTextStyle = logTextStyle;
  }

  ///使用OverlayEntry全局悬浮
  static final OverlayEntry entry = OverlayEntry(
    builder: (context) => Positioned(
      left: _overlayLeft,
      top: _overlayTop,
      child: logOverlayWidget ??
          LogOverlayWidget(
            overlayLeft: _overlayLeft,
            overlayTop: _overlayTop,
            width: _width,
            height: _height,
            logList: logList,
            titleHeight: _titleHeight,
            barColor: _barColor,
            itemColor: _itemColor,
            errorColor: _errorColor,
            needTitle: _needTitle,
            needClean: _needClean,
            logTextStyle: _logTextStyle,
            showListView: showListView,
            onTap: () async {
              showListView = !showListView;
              await _update();
            },
            onDoubleTap: hide,
            onPanUpdate: (DragUpdateDetails detail) async {
              await boundaryConstraint(
                context: context,
                detail: detail,
              );
            },
          ),
    ),
  );

  ///展示悬浮
  static void show() {
    if (_overlay != null) {
      if (!isShow) {
        _overlay!.insert(entry);
      }
    }
  }

  ///隐藏悬浮
  static void hide() {
    if (isShow) {
      entry.remove();
    }
  }

  ///更新悬浮
  static Future<void> _update() async {
    if (isShow) {
      if (_overlay != null) {
        // ignore: invalid_use_of_protected_member
        _overlay!.setState(() {});
      }
    }
  }

  ///边界约束
  static Future<void> boundaryConstraint({
    required BuildContext context,
    required DragUpdateDetails detail,
  }) async {
    if (_overlay != null) {
      //屏幕约束
      final left = _overlayLeft;
      final right = _width + left;
      final top = _overlayTop;
      final bottom = (showListView ? _height : _titleHeight) + top;
      if (left < _mediaQueryData.padding.left) {
        _overlayLeft = _mediaQueryData.padding.left;
      } else if (right >
          (_mediaQueryData.size.width - _mediaQueryData.padding.right)) {
        _overlayLeft =
            _mediaQueryData.size.width - _width - _mediaQueryData.padding.right;
      } else if (top < _mediaQueryData.padding.top) {
        _overlayTop = _mediaQueryData.padding.top;
      } else if (bottom >
          (_mediaQueryData.size.height - _mediaQueryData.padding.bottom)) {
        _overlayTop = _mediaQueryData.size.height -
            (showListView ? _height : _titleHeight) -
            _mediaQueryData.padding.bottom;
      } else {
        _overlayLeft += detail.delta.dx;
        _overlayTop += detail.delta.dy;
      }
      await _update();
    }
  }

  ///增加新日志
  static void addLog({
    required bool isCore,
    required List<String> content,
  }) async {
    logList.add(
      LogOverlayModel(
        isError: isCore,
        content: content,
      ),
    );
    await _update();
  }

  ///清空日志
  static void clearLog() async {
    logList.clear();
    await _update();
  }

  ///上次拖动的时间戳
  static int _lastPanTime = 0;

  ///拖动的次数
  static int _panTimes = 0;

  ///在1秒内朝屏幕左上角拖动3次触发悬浮窗
  static void showOverlayPan(DragEndDetails details) {
    if (details.velocity.pixelsPerSecond.dx < 0 &&
        details.velocity.pixelsPerSecond.dy < 0) {
      final nowTime = DateTime.now();
      if (nowTime
              .difference(DateTime.fromMillisecondsSinceEpoch(_lastPanTime))
              .inMilliseconds <
          1000) {
        if (_panTimes == 3) {
          _panTimes = 0;
          FlutterLogOverlay.show();
        } else {
          _panTimes++;
        }
      } else {
        _lastPanTime = nowTime.millisecondsSinceEpoch;
        _panTimes = 0;
      }
    }
  }
}


lib/src/log_overlay_widget.dart
import 'package:flutter/material.dart';
import 'package:flutter_log_overlay/flutter_log_overlay.dart';

class LogOverlayWidget extends StatelessWidget {
  ///控制弹窗的左上角的x轴坐标(相对于屏幕)
  final double overlayLeft;

  ///控制弹窗的左上角的yx轴坐标(相对于屏幕)
  final double overlayTop;

  ///窗口的宽度
  final double width;

  ///窗口的高度
  final double height;

  ///日志列表
  final List<LogOverlayModel> logList;

  ///标题的高度
  final double titleHeight;

  ///窗口的背景颜色
  final Color? backgroundColor;

  ///标题的背景颜色
  final Color? barColor;

  ///消息的背景颜色
  final Color? itemColor;

  ///error消息的背景颜色
  final Color? errorColor;

  ///日志的TextStyle
  final TextStyle? logTextStyle;

  ///需要标题
  final bool needTitle;

  ///需要清除按钮
  final bool needClean;

  ///是否展开ListView
  final bool showListView;

  ///单击展开或收起ListView
  final void Function()? onTap;

  ///双击隐藏ListView
  final void Function()? onDoubleTap;

  ///拖动改变位置
  final void Function(DragUpdateDetails)? onPanUpdate;

  LogOverlayWidget({
    Key? key,
    required this.overlayLeft,
    required this.overlayTop,
    required this.width,
    required this.height,
    required this.logList,
    required this.titleHeight,
    this.backgroundColor,
    this.barColor,
    this.itemColor,
    this.errorColor,
    this.logTextStyle,
    this.needTitle = true,
    this.needClean = true,
    this.showListView = true,
    this.onTap,
    this.onDoubleTap,
    this.onPanUpdate,
  }) : super(key: key);

  ///控制scrollbar和ListView的滚动控制
  final ScrollController _scrollController = ScrollController();

  @override
  Widget build(BuildContext context) {
    final currentContext = MediaQuery.of(context);
    return Container(
      width: width,
      height: showListView ? height : titleHeight,
      constraints: BoxConstraints(
        minHeight: needTitle ? titleHeight : 0.0,
        maxWidth: currentContext.size.width -
            currentContext.padding.left -
            currentContext.padding.right,
        maxHeight: currentContext.size.height -
            currentContext.padding.top -
            currentContext.padding.bottom,
      ),
      child: Column(
        children: [
          //标题显示
          if (needTitle)
            Material(
              color: Colors.transparent,
              child: buildBar(
                color: barColor ?? Colors.yellowAccent,
              ),
            ),

          //堆栈信息列表
          if (showListView)
            buildListView(
              context: context,
              color: itemColor ?? Colors.greenAccent,
              errorColor: errorColor ?? Colors.redAccent,
            ),
        ],
      ),
    );
  }

  ///标题
  Widget buildBar({
    Color? color,
  }) {
    return Container(
      width: width,
      height: titleHeight,
      color: color,
      alignment: Alignment.centerLeft,
      padding: const EdgeInsets.symmetric(
        horizontal: 5.0,
        vertical: 10.0,
      ),
      child: Row(
        mainAxisAlignment: MainAxisAlignment.spaceBetween,
        children: [
          Expanded(
            child: GestureDetector(
              behavior: HitTestBehavior.translucent,
              onTap: onTap,
              onDoubleTap: onDoubleTap,
              onPanUpdate: onPanUpdate,
              child: FittedBox(
                fit: BoxFit.fitWidth,
                child: Text(
                  "调试控制台(${showListView ? '展开' : '收起'})\n"
                  "当前信息数:${logList.length}\n"
                  "(此区域可以拖动 单击展开或收起 双击隐藏)",
                ),
              ),
            ),
          ),
          if (needClean)
            GestureDetector(
              behavior: HitTestBehavior.translucent,
              onTap: () => FlutterLogOverlay.clearLog(),
              child: const Padding(
                padding: EdgeInsets.symmetric(horizontal: 10.0),
                child: Icon(Icons.cleaning_services),
              ),
            ),
        ],
      ),
    );
  }

  ///ListView
  Widget buildListView({
    required BuildContext context,
    Color? color,
    Color? errorColor,
  }) {
    final mediaQueryFromContext = MediaQuery.of(context);
    return Container(
      width: width,
      height: height - titleHeight,
      constraints: BoxConstraints(
        maxWidth: mediaQueryFromContext.size.width -
            mediaQueryFromContext.padding.left -
            mediaQueryFromContext.padding.right,
        maxHeight: mediaQueryFromContext.size.height -
            mediaQueryFromContext.padding.top -
            mediaQueryFromContext.padding.bottom,
      ),
      child: MediaQuery.removePadding(
        //移除ListView顶部padding
        removeTop: true,
        context: context,
        child: Scrollbar(
          controller: _scrollController,
          radius: const Radius.circular(9),
          thickness: 4.0,
          // 总是显示滚动条
          thumbVisibility: true,
          child: ListView(
            controller: _scrollController,
            scrollDirection: Axis.vertical,
            reverse: true,
            children: logList.reversed
                .map(
                  (e) => Padding(
                    padding: const EdgeInsets.symmetric(vertical: 1.0),
                    child: Material(
                      color: Colors.transparent,
                      child: Container(
                        color: e.isError ? errorColor : color,
                        alignment: Alignment.centerLeft,
                        padding: const EdgeInsets.symmetric(
                          horizontal: 5.0,
                          vertical: 10.0,
                        ),
                        child: SelectableText(
                          e.content
                              .reduce((a, b) => "${a.trim()}\n${b.trim()}"),
                          style:
                              logTextStyle ?? const TextStyle(fontSize: 10.0),
                        ),
                      ),
                    ),
                  ),
                )
                .toList(),
          ),
        ),
      ),
    );
  }
}


example/lib/main.dart
import 'package:flutter/material.dart';
import 'package:flutter_log_overlay/flutter_log_overlay.dart';
import 'package:logger/logger.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return GestureDetector(
      onPanEnd: FlutterLogOverlay.showOverlayPan,
      behavior: HitTestBehavior.translucent,
      child: const MaterialApp(
        debugShowCheckedModeBanner: false,
        home: FirstPgae(),
      ),
    );
  }
}

///first page
class FirstPgae extends StatefulWidget {
  const FirstPgae({
    super.key,
  });

  @override
  State<FirstPgae> createState() => _FirstPgaeState();
}

class _FirstPgaeState extends State<FirstPgae> {
  int _counter = 0;
  void _incrementCounter() {
    sLog.i('first page add:$_counter');
    _counter++;
  }

  @override
  void initState() {
    super.initState();
    WidgetsBinding.instance.addPostFrameCallback((_) {
      FlutterLogOverlay.init(context: context);
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('first page'),
        actions: [
          IconButton(
            onPressed: () {
              Navigator.of(context).push(
                MaterialPageRoute(
                  builder: (context) => const SecondPage(),
                ),
              );
            },
            icon: const Icon(Icons.navigate_next),
          ),
        ],
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: const [
            ElevatedButton(
              onPressed: FlutterLogOverlay.show,
              child: Text('show'),
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ),
    );
  }
}

///second page
class SecondPage extends StatefulWidget {
  const SecondPage({super.key});

  @override
  State<SecondPage> createState() => _SecondPageState();
}

class _SecondPageState extends State<SecondPage> {
  int _counter = 0;
  void _incrementCounter() {
    sLog.i('second page add:$_counter');
    _counter++;
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: const Text('second page'),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: const [
            ElevatedButton(
              onPressed: FlutterLogOverlay.show,
              child: Text('show'),
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: const Icon(Icons.add),
      ),
    );
  }
}

final sLog = MyLogger(
  printer: PrettyPrinter(
    methodCount: 0,
    errorMethodCount: 0,
    printEmojis: false,
    printTime: true,
  ),
);

///类方法重构，增加FlutterLogOverlay.addLog打印日志
class MyLogger extends Logger {
  LogFilter? filter;
  LogPrinter? printer;
  LogOutput? output;
  Level? level;

  MyLogger({
    this.filter,
    this.printer,
    this.output,
    this.level,
  }) : super(
          filter: filter,
          printer: printer,
          output: output,
          level: level,
        );

  @override
  void log(
    Level level,
    dynamic message, [
    dynamic error,
    StackTrace? stackTrace,
  ]) {
    _startTime = DateTime.now();
    FlutterLogOverlay.addLog(
      isCore: level == Level.error,
      content: [
        getTime(),
        "--------------------",
        ((printer ?? PrettyPrinter()) as PrettyPrinter)
            .stringifyMessage(message),
      ],
    );
    super.log(
      level,
      message,
      error != null || stackTrace != null
          ? [
              error,
              stackTrace,
            ]
          : null,
    );
  }

  DateTime? _startTime;

  String getTime() {
    var now = DateTime.now();
    var year = _fourDigits(now.year);
    var month = _twoDigits(now.month);
    var day = _twoDigits(now.day);
    var h = _twoDigits(now.hour);
    var min = _twoDigits(now.minute);
    var sec = _twoDigits(now.second);
    var ms = _threeDigits(now.millisecond);
    var timeSinceStart = now.difference(_startTime!).toString();
    return '$year-$month-$day $h:$min:$sec.$ms (+$timeSinceStart)';
  }

  String _fourDigits(int n) {
    if (n >= 1000) return '$n';
    if (n >= 100) return '0$n';
    if (n >= 10) return '00$n';
    return '000$n';
  }

  String _threeDigits(int n) {
    if (n >= 100) return '$n';
    if (n >= 10) return '0$n';
    return '00$n';
  }

  String _twoDigits(int n) {
    if (n >= 10) return '$n';
    return '0$n';
  }
}
